File: plc4j/transports/udp/src/main/java/org/apache/plc4x/java/transport/udp/UdpTransportInstance.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.transport.udp;

import org.apache.plc4x.java.spi.transports.api.AsyncTransportInstance;
import org.apache.plc4x.java.spi.transports.api.RingBuffer;
import org.apache.plc4x.java.spi.transports.api.exceptions.TransportException;
import org.apache.plc4x.java.transport.udp.config.UdpTransportConfiguration;
import org.apache.plc4x.java.utils.auditlog.api.AuditLog;
import org.apache.plc4x.java.utils.auditlog.api.AuditLogEventType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.util.Optional;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

/**
 * Java 21+ optimized UDP transport (using virtual threads) implementation using NIO DatagramChannel with async support.
 * Implements AsyncTransportInstance for event-driven I/O without polling.
 * Supports both dedicated and shared socket modes for optimal resource usage.
 */
public class UdpTransportInstance implements AsyncTransportInstance<UdpTransportConfiguration> {

    private static final Logger LOGGER = LoggerFactory.getLogger(UdpTransportInstance.class);
    private static final int DEFAULT_BUFFER_SIZE = 8192;
    private static final byte[] EMPTY_BYTES = new byte[0];

    private final UdpTransportConfiguration configuration;
    private final SharedUdpSocketManager sharedUdpSocketManager;
    private final DatagramChannel channel;
    private final InetSocketAddress remoteAddress;
    private final SharedUdpSocketManager.SharedSocket sharedSocket; // null if not shared
    private final RingBuffer ringBuffer;
    private final ByteBuffer readBuffer;  // Pre-allocated direct buffer for zero-copy I/O
    private final int maxPacketSize;
    private final Lock readLock = new ReentrantLock();
    private final Lock writeLock = new ReentrantLock();
    private volatile boolean open = true;

    // Async support
    private final Selector selector;
    private volatile Runnable dataListener;
    private volatile Consumer<Throwable> disconnectListener;
    private final Thread selectorThread;

    public UdpTransportInstance(InetSocketAddress remoteAddress, UdpTransportConfiguration configuration, SharedUdpSocketManager sharedUdpSocketManager, AuditLog auditLog) throws TransportException {
        LOGGER.debug("UdpTransportInstance: Java 21 version");
        this.configuration = configuration;
        this.sharedUdpSocketManager = sharedUdpSocketManager;

        // Use the default max packet size if not configured (handles test cases where defaults aren't applied)
        this.maxPacketSize = configuration.maxPacketSize > 0 ? configuration.maxPacketSize : 65507;
        this.ringBuffer = new RingBuffer(DEFAULT_BUFFER_SIZE);
        this.readBuffer = ByteBuffer.allocateDirect(DEFAULT_BUFFER_SIZE);

        try {
            DatagramChannel tempChannel;
            SharedUdpSocketManager.SharedSocket tempSharedSocket = null;

            if (configuration.shareSocket) {
                // Use shared socket-manager
                tempSharedSocket = sharedUdpSocketManager
                    .acquireSocket(configuration.localAddress, configuration.localPort);
                tempChannel = tempSharedSocket.getChannel();

                LOGGER.debug("Using shared UDP socket on {}:{}",
                    configuration.localAddress, configuration.localPort);
            } else {
                // Create a dedicated socket
                tempChannel = DatagramChannel.open();

                // Bind to a local address and port
                SocketAddress bindAddress;
                if (configuration.localAddress != null && !configuration.localAddress.isEmpty()) {
                    bindAddress = new InetSocketAddress(configuration.localAddress, configuration.localPort);
                } else {
                    bindAddress = new InetSocketAddress(configuration.localPort);
                }
                tempChannel.bind(bindAddress);

                LOGGER.debug("Created dedicated UDP socket bound to {}", bindAddress);
            }

            // Configure socket options
            tempChannel.socket().setBroadcast(configuration.broadcast);
            tempChannel.socket().setReuseAddress(configuration.reuseAddress);

            if (configuration.sendBufferSize > 0) {
                tempChannel.socket().setSendBufferSize(configuration.sendBufferSize);
            }
            if (configuration.receiveBufferSize > 0) {
                tempChannel.socket().setReceiveBufferSize(configuration.receiveBufferSize);
            }
            if (configuration.readTimeout > 0) {
                tempChannel.socket().setSoTimeout(configuration.readTimeout);
            }

            if(!tempChannel.isConnected()) {
                tempChannel.connect(remoteAddress);
            }

            // Configure non-blocking mode for NIO selector
            tempChannel.configureBlocking(false);

            // Create a selector for async I/O
            this.selector = Selector.open();
            tempChannel.register(selector, SelectionKey.OP_READ);

            // Start selector thread using virtual thread (Java 21+)
            this.selectorThread = Thread.ofVirtual()
                .name("UDP-Selector-" + remoteAddress.getHostName() + ":" + remoteAddress.getPort())
            .start(this::runSelectorLoop);

            this.channel = tempChannel;
            this.remoteAddress = remoteAddress;
            this.sharedSocket = tempSharedSocket;

            LOGGER.info("Connected to {}:{} with async support", remoteAddress.getHostName(), remoteAddress.getPort());
            auditLog.write(AuditLogEventType.CONNECT, String.format(
                "Connected to: %s:%d with local address: %s:%d",
                remoteAddress.getHostName(), remoteAddress.getPort(),
                getLocalAddress().orElseThrow().getHostName(), getLocalAddress().orElseThrow().getPort()));
        } catch (IOException e) {
            String errorMsg = String.format("Failed to create UDP transport for %s:%d - %s",
                remoteAddress.getHostName(), remoteAddress.getPort(), e.getMessage());
            LOGGER.error(errorMsg, e);
            auditLog.write(AuditLogEventType.ERROR, "Error in constructor: " + errorMsg);
            auditLog.write(AuditLogEventType.ERROR, "Error in constructor: " + e.getMessage());
            throw new TransportException(errorMsg, e);
        }
    }

    @Override
    public UdpTransportConfiguration getConfiguration() {
        return configuration;
    }

    @Override
    public boolean isOpen() {
        return open && channel.isOpen();
    }

    @Override
    public int getNumBytesAvailable() {
        readLock.lock();
        try {
            if (!isOpen()) {
                return 0;
            }

            return ringBuffer.availableForReading();
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public byte[] peekReadableBytes(int numBytes) throws TransportException {
        if (numBytes <= 0) {
            return EMPTY_BYTES;
        }

        readLock.lock();
        try {
            ensureOpen();

            // Fill the ring buffer if necessary
            getNumBytesAvailable();

            if (ringBuffer.availableForReading() < numBytes) {
                throw new TransportException(
                    String.format("Requested %d bytes but only %d available", numBytes, ringBuffer.availableForReading())
                );
            }

            // Peek without consuming
            return ringBuffer.peek(numBytes);
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public byte[] read(int numBytes) throws TransportException {
        if (numBytes <= 0) {
            return EMPTY_BYTES;
        }

        readLock.lock();
        try {
            ensureOpen();

            // Fill the ring buffer if necessary
            getNumBytesAvailable();

            if (ringBuffer.availableForReading() < numBytes) {
                throw new TransportException(
                    String.format("Requested %d bytes but only %d available", numBytes, ringBuffer.availableForReading())
                );
            }

            // Read and consume bytes
            return ringBuffer.read(numBytes);

        } finally {
            readLock.unlock();
        }
    }

    @Override
    public void write(byte[] bytes) throws TransportException {
        if (bytes == null || bytes.length == 0) {
            return;
        }

        if (bytes.length > maxPacketSize) {
            throw new TransportException(
                String.format("Packet size %d exceeds maximum %d", bytes.length, maxPacketSize)
            );
        }

        writeLock.lock();
        try {
            ensureOpen();

            ByteBuffer writeBuffer = ByteBuffer.wrap(bytes);
            int sent = channel.send(writeBuffer, remoteAddress);

            if (sent != bytes.length) {
                throw new TransportException(
                    String.format("Failed to send complete packet: sent %d of %d bytes", sent, bytes.length)
                );
            }

            LOGGER.trace("Sent {} bytes to {}", sent, remoteAddress);

        } catch (IOException e) {
            throw new TransportException("Failed to send UDP packet", e);
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void close() throws TransportException {
        if (!open) {
            return;
        }

        writeLock.lock();
        try {
            readLock.lock();
            try {
                open = false;

                // Wake up selector
                selector.wakeup();

                if (sharedSocket != null) {
                    // Release shared socket (may not close if other instances use it)
                    sharedUdpSocketManager.releaseSocket(sharedSocket);
                    LOGGER.debug("Released shared UDP socket");
                } else {
                    // Close channel and selector
                    channel.close();
                    selector.close();

                    // Wait for selector thread to finish
                    if (selectorThread != null) {
                        try {
                            selectorThread.join(1000);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    }

                    LOGGER.debug("Closed dedicated UDP socket");
                }

            } catch (IOException e) {
                throw new TransportException("Failed to close UDP transport", e);
            } finally {
                readLock.unlock();
            }
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Receives a packet from the channel and stores it in the ring buffer.
     * Filters packets to only accept from the connected remote address.
     */
    private void receivePacket() throws TransportException {
        try {
            readBuffer.clear();
            SocketAddress source = channel.receive(readBuffer);

            if (source == null) {
                // No packet received (non-blocking mode)
                return;
            }

            readBuffer.flip();

            // Filter packets: only accept from our connected remote address
            if (shouldAcceptPacket(source)) {
                // Copy to ring buffer
                byte[] data = new byte[readBuffer.remaining()];
                readBuffer.get(data);
                ringBuffer.write(data);

                LOGGER.trace("Received packet of {} bytes from {}", data.length, source);
            } else {
                LOGGER.trace("Ignored packet from {} (expected {})", source, remoteAddress);
            }

        } catch (IOException e) {
            throw new TransportException("Failed to receive UDP packet", e);
        }
    }

    /**
     * Checks if a packet from the given source should be accepted.
     * When using shared sockets, multiple instances may receive packets for different remote addresses.
     */
    private boolean shouldAcceptPacket(SocketAddress source) {
        if (!(source instanceof InetSocketAddress inetSource)) {
            return false;
        }

        // Accept if from our remote address
        return inetSource.getAddress().equals(remoteAddress.getAddress()) &&
               inetSource.getPort() == remoteAddress.getPort();
    }

    /**
     * Ensures the transport is still open, throws exception otherwise.
     */
    private void ensureOpen() throws TransportException {
        if (!isOpen()) {
            throw new TransportException("Transport is closed");
        }
    }

    public Optional<InetSocketAddress> getLocalAddress() {
        if(channel != null) {
            return Optional.of((InetSocketAddress) channel.socket().getLocalSocketAddress());
        }
        return Optional.empty();
    }

    // ========== AsyncTransportInstance Implementation ==========

    @Override
    public void registerDataListener(Runnable listener) {
        this.dataListener = listener;
        LOGGER.debug("Data listener registered");
    }

    @Override
    public void removeDataListener() {
        this.dataListener = null;
        LOGGER.debug("Data listener removed");
    }

    @Override
    public void registerDisconnectListener(Consumer<Throwable> listener) {
        this.disconnectListener = listener;
        LOGGER.debug("Disconnect listener registered");
    }

    @Override
    public void removeDisconnectListener() {
        this.disconnectListener = null;
        LOGGER.debug("Disconnect listener removed");
    }

    /**
     * Notifies the disconnect listener if one is registered.
     *
     * @param cause the exception that caused the disconnect, or null for graceful close
     */
    private void notifyDisconnect(Throwable cause) {
        Consumer<Throwable> listener = disconnectListener;
        if (listener != null) {
            try {
                listener.accept(cause);
            } catch (Exception e) {
                LOGGER.error("Error in disconnect listener", e);
            }
        }
    }

    /**
     * Selector loop that runs in a virtual thread and notifies listeners when data arrives.
     * This is the core of the async implementation - no polling needed in the driver!
     */
    private void runSelectorLoop() {
        LOGGER.debug("UDP selector loop started");

        try {
            runSelectorLoopInternal();
        } catch (Throwable t) {
            // This loop is the only thing reading the channel. If it ever exits abnormally
            // there is nobody left to notice, so the transport must stop claiming to be open
            // instead of leaving callers waiting on a connection that can no longer receive.
            LOGGER.error("UDP selector loop failed", t);
            if (open) {
                open = false;
                notifyDisconnect(t);
            }
        }

        LOGGER.debug("UDP selector loop stopped");
    }

    private void runSelectorLoopInternal() {
        while (open && !Thread.currentThread().isInterrupted()) {
            try {
                // Block until events are available (no CPU waste!)
                int readyChannels = selector.select();

                if (readyChannels == 0) {
                    continue;
                }

                var selectedKeys = selector.selectedKeys();
                var iterator = selectedKeys.iterator();

                while (iterator.hasNext()) {
                    SelectionKey key = iterator.next();
                    iterator.remove();

                    if (!key.isValid()) {
                        continue;
                    }

                    if (key.isReadable()) {
                        // Data available - read it into the ring buffer
                        readLock.lock();
                        try {
                            receivePacket(); // Existing method

                            // Notify the listener that data is available
                            Runnable listener = dataListener;
                            if (listener != null) {
                                listener.run();
                            }
                        } catch (TransportException e) {
                            if (open) {
                                LOGGER.error("Error receiving packet", e);
                            }
                        } catch (RuntimeException e) {
                            // Any host can send a datagram to a UDP port, so one we cannot
                            // handle costs us this receive cycle and nothing more.
                            if (open) {
                                LOGGER.error("Error handling received packet", e);
                            }
                        } finally {
                            readLock.unlock();
                        }
                    }
                }

            } catch (IOException e) {
                if (open) {
                    LOGGER.error("Error in selector loop", e);
                    open = false;
                    notifyDisconnect(e);
                }
                break;
            }
        }
    }

}

File: plc4net/transports/udp/UdpTransportInstance.cs
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//

using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using org.apache.plc4net.spi.generation;
using org.apache.plc4net.spi.transports;

namespace org.apache.plc4net.transports.udp
{
    /// <summary>
    /// UDP transport: a bound socket, connected to one remote endpoint, with a
    /// background receive loop feeding a ring buffer.
    /// </summary>
    /// <remarks>
    /// Each <see cref="Write"/> sends one datagram; each datagram received is appended
    /// to the buffer whole, in arrival order. There is no stream, so a codec that
    /// wants message boundaries relies on the protocol's own length field
    /// (KNXnet/IP's total-length header) — the same as over TCP.
    /// </remarks>
    public class UdpTransportInstance : BaseTransportInstance, IAsyncTransportInstance
    {
        private readonly Socket _socket;
        private readonly RingBuffer _ringBuffer;
        private readonly object _readLock = new object();
        private readonly object _writeLock = new object();
        private readonly CancellationTokenSource _shutdown = new CancellationTokenSource();
        private readonly Task _receiveLoop;

        private const int MaxConsecutiveTransientErrors = 20;
        private const int TransientBackoffMs = 50;

        private int _open = 1;
        private int _released;

        private volatile Action _dataListener;
        private volatile Action<Exception> _disconnectListener;
        private int _listenerThreadId;

        public UdpTransportInstance(IPEndPoint remoteAddress, UdpTransportConfiguration configuration)
            : base(configuration)
        {
            if (remoteAddress == null)
            {
                throw new ArgumentNullException(nameof(remoteAddress));
            }

            _ringBuffer = new RingBuffer(configuration.ReceiveBufferSize);

            Socket socket = null;
            try
            {
                socket = new Socket(remoteAddress.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
                if (configuration.Broadcast)
                {
                    socket.EnableBroadcast = true;
                }

                var bindAddress = string.IsNullOrEmpty(configuration.LocalAddress)
                    ? (remoteAddress.AddressFamily == AddressFamily.InterNetworkV6
                        ? IPAddress.IPv6Any
                        : IPAddress.Any)
                    : IPAddress.Parse(configuration.LocalAddress);
                socket.Bind(new IPEndPoint(bindAddress, configuration.LocalPort));

                // "Connecting" a datagram socket fixes the destination for Send and
                // filters inbound datagrams to that peer - exactly the KNXnet/IP model
                // (one gateway per connection).
                socket.Connect(remoteAddress);

                _socket = socket;
                RemoteAddress = remoteAddress;
                LocalAddress = socket.LocalEndPoint as IPEndPoint;
            }
            catch (Exception e)
            {
                socket?.Dispose();
                throw new TransportException(
                    $"Failed to open a UDP socket to {remoteAddress.Address}:{remoteAddress.Port} - {e.Message}", e);
            }

            _receiveLoop = Task.Run(RunReceiveLoopAsync);
        }

        public IPEndPoint RemoteAddress { get; }

        /// <summary>The local endpoint the socket bound to. KNXnet/IP advertises this
        /// to the gateway in its HPAI fields, so the gateway knows where to reply.</summary>
        public IPEndPoint LocalAddress { get; }

        public override bool IsOpen => Volatile.Read(ref _open) == 1;

        internal bool ResourcesReleased => Volatile.Read(ref _released) == 1;

        /// <summary>Optional diagnostic output: written and read bytes as annotated hex.</summary>
        public TextWriter DiagnosticOutput { get; set; }

        public override int GetNumBytesAvailable()
        {
            lock (_readLock)
            {
                return _ringBuffer.AvailableForReading;
            }
        }

        public override byte[] PeekReadableBytes(int numBytes)
        {
            if (numBytes <= 0)
            {
                return Array.Empty<byte>();
            }
            lock (_readLock)
            {
                if (_ringBuffer.AvailableForReading < numBytes)
                {
                    throw new TransportException(
                        $"Requested {numBytes} bytes but only {_ringBuffer.AvailableForReading} available");
                }
                return _ringBuffer.Peek(numBytes);
            }
        }

        public override byte[] Read(int numBytes)
        {
            if (numBytes <= 0)
            {
                return Array.Empty<byte>();
            }
            byte[] result;
            lock (_readLock)
            {
                if (_ringBuffer.AvailableForReading < numBytes)
                {
                    throw new TransportException(
                        $"Requested {numBytes} bytes but only {_ringBuffer.AvailableForReading} available");
                }
                result = _ringBuffer.Read(numBytes);
            }
            WriteHex("UDP RECV", result);
            return result;
        }

        public override void Write(byte[] bytes)
        {
            if (bytes == null || bytes.Length == 0)
            {
                return;
            }
            lock (_writeLock)
            {
                EnsureOpen();
                try
                {
                    // One datagram per call. A short Send on a datagram socket means the
                    // datagram did not fit; treat it as an error rather than fragmenting.
                    var sent = _socket.Send(bytes, 0, bytes.Length, SocketFlags.None);
                    if (sent != bytes.Length)
                    {
                        throw new TransportException(
                            $"UDP datagram truncated: sent {sent} of {bytes.Length} bytes.");
                    }
                    WriteHex("UDP SEND", bytes);
                }
                catch (ObjectDisposedException) when (Volatile.Read(ref _open) == 0)
                {
                    // A concurrent Close() disposed the socket mid-write: normal shutdown.
                }
                catch (SocketException e)
                {
                    throw new TransportException("Failed to send datagram", e);
                }
            }
        }

        public override void Close()
        {
            Volatile.Write(ref _open, 0);
            ReleaseResources(fromReceiveLoop: false);
        }

        private void ReleaseResources(bool fromReceiveLoop)
        {
            if (Interlocked.CompareExchange(ref _released, 1, 0) != 0)
            {
                return;
            }

            _shutdown.Cancel();
            try
            {
                _socket.Dispose();
            }
            catch (ObjectDisposedException)
            {
            }
            finally
            {
                var insideLoop = fromReceiveLoop
                    || Volatile.Read(ref _listenerThreadId) == Environment.CurrentManagedThreadId;
                if (!insideLoop && _receiveLoop != null && !_receiveLoop.IsCompleted)
                {
                    _receiveLoop.Wait(TimeSpan.FromSeconds(1));
                }
                _shutdown.Dispose();
            }
        }

        public void RegisterDataListener(Action listener)
        {
            _dataListener = listener;

            // A datagram may have arrived and been buffered before the listener was
            // registered; deliver one wake-up so the codec drains what is waiting.
            bool hasBufferedData;
            lock (_readLock)
            {
                hasBufferedData = _ringBuffer.AvailableForReading > 0;
            }
            if (listener != null && hasBufferedData)
            {
                NotifyData();
            }
        }

        public void RemoveDataListener() => _dataListener = null;

        public void RegisterDisconnectListener(Action<Exception> listener) => _disconnectListener = listener;

        public void RemoveDisconnectListener() => _disconnectListener = null;

        private void NotifyDisconnect(Exception cause)
        {
            var listener = _disconnectListener;
            if (listener == null)
            {
                return;
            }
            Volatile.Write(ref _listenerThreadId, Environment.CurrentManagedThreadId);
            try
            {
                listener(cause);
            }
            catch
            {
                // A misbehaving listener must not take down the receive loop.
            }
            finally
            {
                Volatile.Write(ref _listenerThreadId, 0);
            }
        }

        private void NotifyData()
        {
            var listener = _dataListener;
            if (listener == null)
            {
                return;
            }
            Volatile.Write(ref _listenerThreadId, Environment.CurrentManagedThreadId);
            try
            {
                listener();
            }
            catch
            {
                // Same reasoning as NotifyDisconnect.
            }
            finally
            {
                Volatile.Write(ref _listenerThreadId, 0);
            }
        }

        private async Task RunReceiveLoopAsync()
        {
            var datagram = new byte[65535];
            var transientErrors = 0;
            try
            {
                while (Volatile.Read(ref _open) == 1)
                {
                    int received;
                    try
                    {
                        received = await _socket
                            .ReceiveAsync(new ArraySegment<byte>(datagram), SocketFlags.None)
                            .ConfigureAwait(false);
                        transientErrors = 0;
                    }
                    catch (SocketException e) when (IsTransientReceiveError(e.SocketErrorCode))
                    {
                        // ICMP port/host-unreachable from the peer (ConnectionReset on
                        // Windows, ConnectionRefused on Linux), or an oversized datagram
                        // (MessageSize). The gateway may just not be up yet; back off and
                        // keep listening rather than tearing the transport down. Escalate
                        // only if the errors do not stop.
                        if (++transientErrors > MaxConsecutiveTransientErrors)
                        {
                            throw;
                        }
                        await Task.Delay(TransientBackoffMs, _shutdown.Token).ConfigureAwait(false);
                        continue;
                    }

                    if (received <= 0)
                    {
                        continue;
                    }

                    var accepted = false;
                    lock (_readLock)
                    {
                        if (received <= _ringBuffer.Capacity)
                        {
                            if (_ringBuffer.RemainingForWriting < received)
                            {
                                // The codec has not kept up. Discard everything buffered
                                // rather than drop bytes from the middle of a frame - a
                                // partial frame left behind desyncs the length-prefixed
                                // codec permanently. The next datagram starts clean.
                                _ringBuffer.Clear();
                            }
                            _ringBuffer.Write(datagram, 0, received);
                            accepted = true;
                        }
                    }

                    if (accepted)
                    {
                        NotifyData();
                    }
                    // else: a datagram larger than the whole buffer can never be
                    // delivered intact; drop it and stay alive.
                }
            }
            catch (OperationCanceledException)
            {
                // Close() cancelled us.
            }
            catch (ObjectDisposedException)
            {
                // Close() disposed the socket while we were awaiting a receive.
            }
            catch (Exception e)
            {
                if (Volatile.Read(ref _open) == 1)
                {
                    Volatile.Write(ref _open, 0);
                    NotifyDisconnect(e);
                }
            }
            finally
            {
                Volatile.Write(ref _open, 0);
                ReleaseResources(fromReceiveLoop: true);
            }
        }

        private static bool IsTransientReceiveError(SocketError code) =>
            code == SocketError.ConnectionReset       // Windows: ICMP port-unreachable
            || code == SocketError.ConnectionRefused  // Linux: ICMP port-unreachable
            || code == SocketError.MessageSize        // datagram larger than the buffer
            || code == SocketError.HostUnreachable
            || code == SocketError.NetworkUnreachable;

        private void WriteHex(string label, byte[] bytes)
        {
            var diag = DiagnosticOutput;
            if (diag == null)
            {
                return;
            }
            diag.WriteLine($"┌ {label} ({bytes.Length} bytes)");
            for (var i = 0; i < bytes.Length; i += 16)
            {
                diag.Write("│ ");
                for (var j = i; j < Math.Min(i + 16, bytes.Length); j++)
                {
                    diag.Write($"{bytes[j]:X2} ");
                }
                diag.WriteLine();
            }
            diag.WriteLine("└");
            diag.Flush();
        }
    }
}
